db/dohsql.c

```diff
@@ -1361,6 +1381,16 @@ void dohsql_wait_for_master(sqlite3_stmt *stmt, struct sqlclntstate *clnt)
         while (conn->status == DOH_RUNNING && queue_count(conn->que) > 0) {
             Pthread_mutex_unlock(&conn->mtx);
             poll(NULL, 0, 10);
+            if (bdb_lock_desired(thedb->bdb_env)) {
```

Comment:
NOTE: this should be safe imho since we are calling this from parallel worker thread when its work is done; it waits for master to consume its row from que before freeing the resources allocated and terminating.